larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "SlidingLinearFitWindowLarge", m_slidingLinearFitWindowLarge));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddStraightLineLength", m_addStraightLineLength));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddDiffWithStraightLineMean", m_addDiffWithStraightLineMean));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddDiffWithStraightLineSigma", m_addDiffWithStraightLineSigma));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddDTDLWidth", m_addDTDLWidth));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddMaxFitGapLength", m_addMaxFitGapLength));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddRMSLinearFit", m_addRMSLinearFit));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "RatioVariables", m_ratioVariables));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ThreeDVertexDistanceFeatureTool::ThreeDVertexDistanceFeatureTool() :
+    m_ratioVariables(false),
+    m_slidingLinearFitWindow(10000)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeDVertexDistanceFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    //find the neutrino vertex
+    const VertexList *pVertexList = nullptr;
+    (void) PandoraContentApi::GetCurrentList(*pAlgorithm, pVertexList);
+
+    if (pVertexList->empty() || (pVertexList->size() != 1) || (VERTEX_3D != pVertexList->front()->GetVertexType()))
+    {
+        throw StatusCodeException(STATUS_CODE_FAILURE);
+    }
+
+    const Vertex *const nuVertex(pVertexList->front());
+    //find the particle vertex
+    const Vertex *const pVertex = LArPfoHelper::GetVertex(pInputPfo);
```

Comment:
Presumably there's no way that the input pfo is a neutrino here (with the neutrino vertex)?